File: wpimath/src/test/java/edu/wpi/first/math/controller/serde/DifferentialDriveWheelVoltagesSerdeTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.math.controller.serde;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;

import edu.wpi.first.math.controller.DifferentialDriveWheelVoltages;
import edu.wpi.first.math.proto.Controller.ProtobufDifferentialDriveWheelVoltages;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import org.junit.jupiter.api.Test;

public class DifferentialDriveWheelVoltagesSerdeTest {
  private static final DifferentialDriveWheelVoltages DATA =
      new DifferentialDriveWheelVoltages(1.91, 2.29);
  private static final byte[] STRUCT_BUFFER =
      new byte[] {-113, -62, -11, 40, 92, -113, -2, 63, 82, -72, 30, -123, -21, 81, 2, 64};

  @Test
  void testStructPack() {
    ByteBuffer buffer = ByteBuffer.allocate(DifferentialDriveWheelVoltages.struct.getSize());
    buffer.order(ByteOrder.LITTLE_ENDIAN);
    DifferentialDriveWheelVoltages.struct.pack(buffer, DATA);

    byte[] actual = buffer.array();
    assertArrayEquals(actual, STRUCT_BUFFER);
  }

  @Test
  void testStructUnpack() {
    ByteBuffer buffer = ByteBuffer.wrap(STRUCT_BUFFER);
    buffer.order(ByteOrder.LITTLE_ENDIAN);

    DifferentialDriveWheelVoltages data = DifferentialDriveWheelVoltages.struct.unpack(buffer);
    assertEquals(DATA.left, data.left);
    assertEquals(DATA.right, data.right);
  }

  @Test
  void testProtoPack() {
    ProtobufDifferentialDriveWheelVoltages proto =
        DifferentialDriveWheelVoltages.proto.createMessage();
    DifferentialDriveWheelVoltages.proto.pack(proto, DATA);

    assertEquals(DATA.left, proto.getLeft());
    assertEquals(DATA.right, proto.getRight());
  }

  @Test
  void testProtoUnpack() {
    ProtobufDifferentialDriveWheelVoltages proto =
        DifferentialDriveWheelVoltages.proto.createMessage();
    proto.setLeft(DATA.left);
    proto.setRight(DATA.right);

    DifferentialDriveWheelVoltages data = DifferentialDriveWheelVoltages.proto.unpack(proto);
    assertEquals(DATA.left, data.left);
    assertEquals(DATA.right, data.right);
  }
}

File: wpimath/src/test/native/cpp/controller/serde/DifferentialDriveWheelVoltagesSerdeTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <gtest/gtest.h>

#include "frc/geometry/DifferentialDriveWheelVoltages.h"
#include "geometry2d.pb.h"

using namespace frc;

namespace {

using StructType = wpi::Struct<frc::DifferentialDriveWheelVoltages>;
using ProtoType = wpi::Protobuf<frc::DifferentialDriveWheelVoltages>;

constexpr std::array<uint8_t, StructType::kSize> create_test_buffer() {
  std::array<uint8_t, StructType::kSize> output;
  int buffer[] = {
    0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0,
};
  for (size_t idx = 0; idx < StructType::kSize; ++idx) {
    output[idx] = static_cast<uint8_t>(buffer[idx]);
  }
  return output;
}

std::array<uint8_t, StructType::kSize> kExpectedStructBuffer =
    create_test_buffer();

constexpr DifferentialDriveWheelVoltages kExpectedData{};
}  // namespace

TEST(DifferentialDriveWheelVoltagesTest, StructPack) {
  uint8_t buffer[StructType::kSize];
  StructType::Pack(buffer, kExpectedData);

  for (size_t i = 0; i < StructType::kSize; ++i) {
    EXPECT_EQ(kExpectedStructBuffer[i], buffer[i]) << " on byte " << i;
  }
}

TEST(DifferentialDriveWheelVoltagesTest, StructUnpack) {
  DifferentialDriveWheelVoltages unpacked_data = StructType::Unpack(kExpectedStructBuffer);

    EXPECT_EQ(kExpectedData.Left(), unpacked_data.Left());
    EXPECT_EQ(kExpectedData.Right(), unpacked_data.Right());
}


TEST(DifferentialDriveWheelVoltagesTest, ProtobufPack) {
  wpi::proto::ProtobufDifferentialDriveWheelVoltages proto;
  ProtoType::Pack(&proto, kExpectedData);

    EXPECT_EQ(kExpectedData.Left(), proto.left());
    EXPECT_EQ(kExpectedData.Right(), proto.right());
}



TEST(DifferentialDriveWheelVoltagesTest, ProtobufUnpack) {
  wpi::proto::ProtobufDifferentialDriveWheelVoltages proto;
  proto.set_left(kExpectedData.Left());
  proto.set_right(kExpectedData.Right());

  DifferentialDriveWheelVoltages unpacked_data = ProtoType::Unpack(proto);
    EXPECT_EQ(kExpectedData.Left(), unpacked_data.Left());
    EXPECT_EQ(kExpectedData.Right(), unpacked_data.Right());
}